### ISSUE TITLE
Doubts about KL loss

### ISSUE THREAD

**Issue by LianShuaiLong** [lucidrains/DALLE-pytorch#255]
Here are your codes when calculate KL loss:
```
        log_qy = F.log_softmax(logits, dim = -1)
        log_uniform = torch.log(torch.tensor([1. / num_tokens], device = device))
        kl_div = F.kl_div(log_uniform, log_qy, None, None, 'batchmean', log_target = True)
```
but according to pytorch Doc about kl_div:

[kl_div](https://pytorch.org/docs/stable/nn.functional.html?highlight=kl_div#torch.nn.functional.kl_div)

1. log_uniform should has the same shape as log_qy
2. log_uniform and log_qy should  swap position

Is there anything wrong?

**Comment by avihu111** [comment]
I'm joining the doubts about the KL term.
But before that, let me try to address your two cases:
1. The same shape tensor outputs the exact same result
2. I'll try to convince why this is the correct way to write this. 
For simplicity, consider the following 2d distributions: expected=(0.5,0.5) and actual=(0,1).
When this the case, we would like the loss to be high, cause the network doesn't output the expected uniform distribution.
notice that KL(expected|actual)=inf and KL(actual|expected)=log(2)

**Comment by avihu111** [comment]
I tried running with the KL term got a much more sparse distribution than without it: 
![image](https://user-images.githubusercontent.com/39214195/119867603-bb7b1400-bf26-11eb-9978-2ee426e9894a.png)
so I thought that the objective might be slightly off from what we intend. 
As I see it, we try to maintain the codebook distribution uniform. This might be better to do this per example, and not per pixel. I tried comparing the mean softmax probabilities to the uniform distribution. this is the code:
`        logits = rearrange(logits, 'b n h w -> b (h w) n')
        softmax = F.softmax(logits, dim=-1)
        mean_softmax = softmax.mean(dim=1)
        log_qy = torch.log(mean_softmax)
        log_uniform = torch.log(torch.tensor([1. / num_tokens], device = device))
        kl_div = F.kl_div(log_uniform, log_qy, None, None, 'batchmean', log_target = True)
        loss = recon_loss + kl_div_loss_weight * kl_div`

when training with this (using the same kl_div_loss_weight=0.001) I get a more sensible results:
![image](https://user-images.githubusercontent.com/39214195/119868529-d0a47280-bf27-11eb-9c89-5140db0b962f.png)

I'll check how this behaves with higher kl_div_loss_weight values. 
If you think my logic is fault I'd love to hear

**Comment by afiaka87** [comment]
Some of this discussion may be relevant

https://github.com/lucidrains/DALLE-pytorch/issues/250

**Comment by LianShuaiLong** [comment]
> I'm joining the doubts about the KL term.
> But before that, let me try to address your two cases:
> 
> 1. The same shape tensor outputs the exact same result
> 2. I'll try to convince why this is the correct way to write this.
>    For simplicity, consider the following 2d distributions: expected=(0.5,0.5) and actual=(0,1).
>    When this the case, we would like the loss to be high, cause the network doesn't output the expected uniform distribution.
>    notice that KL(expected|actual)=inf and KL(actual|expected)=log(2)

According to F.kl_div api:
`torch.nn.functional.kl_div(input, target, size_average=None, reduce=None, reduction='mean', log_target=False)`


the kl term should write like this:
`kl_div = F.kl_div(log_uniform, log_qy, None, None, 'batchmean', log_target = True)`

since log_uniform is the 'target' and log_qy is the 'input',right?

**Comment by lucidrains** [comment]
@LianShuaiLong I am following the tutorial as put out here in this blog post https://blog.evjang.com/2016/11/tutorial-categorical-variational.html

specifically, this line `kl_tmp = tf.reshape(q_y*(log_q_y-tf.log(1.0/K)),[-1,N,K])`

**Comment by lucidrains** [comment]
@LianShuaiLong 

```python
import torch
import torch.nn.functional as F

eps = 1e-20
uniform_prob = torch.tensor([0.5, 0.5])
prob = torch.tensor([1., 0])

log_prob = torch.log(prob + eps)
log_uniform = torch.log(uniform_prob)

kl_tmp = (prob * (log_prob - log_uniform)).mean()

kl_tmp_fn = F.kl_div(log_uniform, log_prob, log_target = True)
print(kl_tmp, kl_tmp_fn)
```
